### PR TITLE
[BEAM-2530] Compile and run tests on java 11 for PreCommit portability api

### PR DESCRIPTION
**Pull request by pawelpasterz**
This is next step for java 11 test compilation. #11527 enabled this approach.

For task within test should be compiled with java 11 add `-PcompileAndRunTestsWithJava11` [1] and `-Pjava11Home=[path_to_java_11_home]` [2]

[1] adds tasks dependency to run `verifyJavaVersion` before
```
if (project.hasProperty('compileAndRunTestsWithJava11')) {
  project.javaPreCommitPortabilityApi.dependsOn ':sdks:java:testing:test-utils:verifyJavaVersion'
}
```
Task contains 3 JUnit tests that verify:
1. if runtime jvm version is 11
2. if beam bytecode version is 8
3. if test bytecode version is 11

If [1] is not present then verification tests are excluded from tasks of type `Test`.
```
} else {
  allprojects {
    tasks.withType(Test) {
      exclude '**/JvmVerification.class'
    }
  }
}
```

[2] is used in forking options for `compileTestJava` task and as executable for `Test` type tasks.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by pawelpasterz** [comment]
R: @Ardagan @iemejia @kennknowles @mwalenia 
Please, take a look and let me know what do you think

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 1]
```diff
@@ -698,6 +700,21 @@ class BeamModulePlugin implements Plugin<Project> {
...
       }
 
+      if (project.hasProperty("compileAndRunTestsWithJava11")) {
+        def java11Home = project.findProperty("java11Home")
+        project.tasks.compileTestJava {
+          options.fork = true
+          options.forkOptions.javaHome = java11Home as File
+          sourceCompatibility = JavaVersion.VERSION_1_8
+          targetCompatibility = JavaVersion.VERSION_11
+          options.compilerArgs += ['-Xlint:-path']
```
[the remark this answers is not in the document]
Comment (pawelpasterz) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:710]: Sure, thanks for pointing this out

**Comment by pawelpasterz** [comment]
> Can you please squash the commits and rename them to include the valid JIRA name e.g. "[[BEAM-2530](https://issues.apache.org/jira/browse/BEAM-2530)] ..."

Sure, I'll squash commits once PR will have all requested changes implemented

> Just for curiosity, since this only build/run the tests with Java 11. What are the current blockers for the full build with Java 11?

Prior was error prone plugin version. Now, when I try to compile beam with java I get errors like:
```
> Task :model:pipeline:compileJava
/Users/pawel/beam/model/pipeline/build/generated/source/proto/main/grpc/org/apache/beam/model/pipeline/v1/TestStreamServiceGrpc.java:20: error: cannot find symbol
@javax.annotation.Generated(
                 ^
  symbol:   class Generated
  location: package javax.annotation
1 error
```

TBH, I did not spend much time figuring this out since it is not main concern of this PR. I'll definitely try to fix this in the future. Or maybe you have some tips regarding this error?

**Comment by iemejia** [comment]
> Prior was error prone plugin version. Now, when I try to compile beam with java I get errors like:
> 
> ```
> > Task :model:pipeline:compileJava
> /Users/pawel/beam/model/pipeline/build/generated/source/proto/main/grpc/org/apache/beam/model/pipeline/v1/TestStreamServiceGrpc.java:20: error: cannot find symbol
> @javax.annotation.Generated(
>                  ^
>   symbol:   class Generated
>   location: package javax.annotation
> 1 error
> ```
> 
> TBH, I did not spend much time figuring this out since it is not main concern of this PR. I'll definitely try to fix this in the future. Or maybe you have some tips regarding this error?

Yes it is not necessary to do this for this PR. Those errors are because some modules are not part of Java by default anymore and should be included explicitly with `--add-module` and the right module.

**Comment by iemejia** [comment]
@pawelpasterz I am ok with the current version can you please squash and update so we can run the tests and see if everything work as expected before merge.

Also this has probably to be updated once https://issues.apache.org/jira/browse/BEAM-9915 is done.

**Comment by pawelpasterz** [comment]
> @pawelpasterz I am ok with the current version can you please squash and update so we can run the tests and see if everything work as expected before merge.
> 
> Also this has probably to be updated once https://issues.apache.org/jira/browse/BEAM-9915 is done.

Yes, definitely!

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
retest this please

**Comment by mwalenia** [comment]
retest this please

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run JavaPortabilityApiJava11 PreCommit

**Comment by mwalenia** [comment]
Run JavaPortabilityApi PreCommit

**Comment by mwalenia** [comment]
Run JavaPortabilityApi PreCommit

**Comment by mwalenia** [comment]
retest this please

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run JavaPortabilityApiJava11 PreCommit

**Comment by mwalenia** [comment]
Run JavaPortabilityApi PreCommit

**Comment by mwalenia** [comment]
retest this please

**Comment by mwalenia** [comment]
retest this please

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
run seed job

**Comment by mwalenia** [comment]
Run JavaPortabilityApiJava11 PreCommit

**Comment by mwalenia** [comment]
Run JavaPortabilityApi PreCommit

**Comment by mwalenia** [comment]
Run JavaPortabilityApiJava11 PreCommit

**Comment by mwalenia** [comment]
Run JavaPortabilityApi PreCommit

**Comment by iemejia** [comment]
Merged manually to add the missing ticket prefix `[BEAM-2530`. Thanks again!

**Pull request closed without merging**